### ISSUE TITLE
Do not store callbacks/procs in shared memory!

### ISSUE THREAD

**Issue by cheatfate** [nim-lang/Nim#4439]
I have lost 2 days because of this problem, so for every people who wants to store callbacks or procedures in shared memory:

You need to protect your callbacks environment with this function:
``` nimrod
proc protect*(x: pointer): ForeignCell =
```
[gc_common.nim:15](https://github.com/nim-lang/Nim/blob/devel/lib/system/gc_common.nim#L15)

And dispose your callback's environment with this function:
``` nimrod
proc dispose*(x: ForeignCell) =
```
[gc_common.nim:20](https://github.com/nim-lang/Nim/blob/devel/lib/system/gc_common.nim#L20)

To get callback's environment pointer you need this function:
``` nimrod
proc rawEnv*[T: proc](x: T): pointer {.noSideEffect, inline.} =
```
[system.nim:3062](https://github.com/nim-lang/Nim/blob/devel/lib/system.nim#L3062)

If you not protect your callback with **protect()**, GC can free your callback's environment, which stored in **GC memory**, and your callback can cause **SIGSEGV** or wrong arguments/environment values.

**Comment by yglukhov** [comment]
Does not `GC_ref` do  the same thing?

**Comment by cheatfate** [comment]
@yglukhov, yep, but, problems is `rawEnv()` returns **pointer**. And we have only this procs:
``` nimrod
proc GC_ref*[T](x: ref T) {.magic: "GCref", benign.}
proc GC_ref*[T](x: seq[T]) {.magic: "GCref", benign.}
proc GC_ref*(x: string) {.magic: "GCref", benign.}
```
So we need to convert **pointer** to **ref T**, but we dont have **T** for environment variable...

**Comment by yglukhov** [comment]
I remember Araq told me that `ref _anything_` can safely be casted to `ref _anything_else_` and GC will not corrupt it. Given that closure env is a `ref object`, we can safely cast `rawEnv` to `ref object` and `GC_ref` it.
```nim
GC_ref(cast[ref object](rawEnv(myClosure)))
```
@Araq, am I right? If yes, what is the whole purpose of `protect`/`unprotect` then?

**Comment by Araq** [comment]
@yglukhov you're right. The purpose of ``protect`` / ``dispose`` is that it's threadsafe, ``dispose`` can be called from any thread, ``protect`` needs to be called by the owning thread though.

**Comment by andreaferretti** [comment]
Since the purpose of this issue is to warn readers about a possible pitfall, I think the content of this discussion should go in the documentation somewhere

**Comment by cheatfate** [comment]
This should be documented properly somewhere.
